codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/codegen/ValidationGenerator.java

```diff
@@ -565,4 +438,147 @@ private StringBuilder addUTFCheck(
+      final var listValidation = new StringBuilder();
+      listValidation.append(
+        """
+        for _, %s := range %s {
+        """.formatted(LIST_ITEM, dataSource)
+      );
+      listValidation.append(itemValidation);
+      listValidation.append(
+        """
+        }
+        """
+      );
+      validationFuncMap.put(memberShape, listValidation.toString());
+    }
+  }
+
+  private void renderMapShape(final MapShape currentShape, final MemberShape  memberShape, final StringBuilder validationCode, String dataSource) {
+    final var keyMember = currentShape.getKey();
+    final var valueMember = currentShape.getValue();
+    final var keyValidation = new StringBuilder();
+    final var valueValidation = new StringBuilder();
+    renderValidatorForEachShape(model.expectShape(keyMember.getTarget()), keyMember, false, MAP_KEY, keyValidation);
+    renderValidatorForEachShape(model.expectShape(valueMember.getTarget()), valueMember, false, MAP_VALUE, valueValidation);
+    // Put map key or value to _ if no constraints in key or value
+    final var maybeMapKey = keyValidation.isEmpty() ? "_" : MAP_KEY;
+    final var maybeMapValue = valueValidation.isEmpty() ? "_" : MAP_VALUE;
+    // If the validation function is not created and the map shape does have some constraints in its key and value
+    if (!validationFuncMap.containsKey(memberShape) && (!keyValidation.isEmpty() || !valueValidation.isEmpty())) {
+      final var funcName = funcNameGenerator(memberShape, "validate");
+      final var funcInput = dataSource.startsWith("input") ? "" : dataSource;
+      if (!funcInput.isEmpty()) {
+        var inputType = GoCodegenUtils.getType(symbolProvider.toSymbol(currentShape), currentShape);
+        // remove the package name because this code is generated inside smithyTypesNamespace itself
+        inputType =
+          inputType.replace(
+            SmithyNameResolver.smithyTypesNamespace(currentShape).concat("."),
+            ""
+          );
+        validationFuncInputTypeMap.put(memberShape, inputType);
+        dataSource = "Value";
+      }
+      final var funcCall = "input.".concat(funcName).concat("(%s)".formatted(funcInput));
+      validationCode.append(
+        CHECK_AND_RETURN_ERROR.formatted(
+          funcCall, funcCall
+        )
+      );
+      validationFuncMap.put(memberShape, null);
+      final var mapValidation = new StringBuilder();
+      mapValidation.append(
+        """
+        for %s, %s := range %s {
+        """.formatted(maybeMapKey, maybeMapValue, dataSource)
+      );
+      mapValidation.append(keyValidation);
+      mapValidation.append(valueValidation);
+      mapValidation.append(
+        """
+            }
+        """
+      );
+      validationFuncMap.put(memberShape, mapValidation.toString());
+    }
+  }
+  private void renderUnionShape(final UnionShape currentShape, final MemberShape  memberShape, final StringBuilder validationCode, String dataSource) {
```

Comment:
Non blocking (because I think `make format_java` will take care of it):

There should be a white space between `}` and `private void renderUnionShape`

Reply:
Fixed manually. Didn't want formatting changes with this PR.